src/io/dirops.c

```diff
@@ -207,163 +124,82 @@ static const MVMIOOps op_table = {
+        MVM_exception_throw_adhoc(tc, "%s got incorrect kind of handle", msg);
+    return handle;
+}
+
 /* Open a filehandle, returning a handle. */
 MVMObject * MVM_dir_open(MVMThreadContext *tc, MVMString *dirname) {
     MVMOSHandle  * result;
-    MVMIODirIter * const data   = MVM_calloc(1, sizeof(MVMIODirIter));
+    int opendir_error = 0;
+    uv_fs_t req;
     MVMROOT(tc, dirname, {
         result = (MVMOSHandle *)MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTIO);
     });
-#ifdef _WIN32
-    {
-        char *name;
-        int str_len;
-        wchar_t *wname;
-        wchar_t *dir_name;
-
-        name  = MVM_string_utf8_c8_encode_C_string(tc, dirname);
-        wname = UTF8ToUnicode(name);
-        MVM_free(name);
-
-        str_len = wcslen(wname);
-
-        if (str_len > MAX_PATH - 2) { // the length of later appended '\*' is 2
-            wchar_t  abs_dirname[4096]; /* 4096 should be enough for absolute path */
-            wchar_t *lpp_part;
-
-            /* You cannot use the "\\?\" prefix with a relative path,
-             * relative paths are always limited to a total of MAX_PATH characters.
-             * see http://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx */
-            if (!GetFullPathNameW(wname, 4096, abs_dirname, &lpp_part)) {
-                MVM_free(data);
-                MVM_free(wname);
-                MVM_exception_throw_adhoc(tc, "Directory path is wrong: %lu", GetLastError());
-            }
-            MVM_free(wname);
-
-            str_len  = wcslen(abs_dirname);
-            dir_name = (wchar_t *)MVM_malloc((str_len + 7) * sizeof(wchar_t));
-            wcscpy(dir_name, L"\\\\?\\");
-            wcscat(dir_name, abs_dirname);
-        } else {
-            dir_name = (wchar_t *)MVM_malloc((str_len + 3) * sizeof(wchar_t));
-            wcscpy(dir_name, wname);
-            MVM_free(wname);
-        }
-
-        wcscat(dir_name, L"\\*");     /* Three characters are for the "\*" plus NULL appended.
-                                       * see http://msdn.microsoft.com/en-us/library/windows/desktop/aa365200%28v=vs.85%29.aspx */
 
-        data->dir_name   = dir_name;
-        data->dir_handle = INVALID_HANDLE_VALUE;
-    }
-
-#else
-    {
-        char * const dir_name = MVM_string_utf8_c8_encode_C_string(tc, dirname);
-        DIR * const dir_handle = opendir(dir_name);
-        int opendir_error = errno;
-        MVM_free(dir_name);
-
-        if (!dir_handle) {
-            MVM_free(data);
-            MVM_exception_throw_adhoc(tc, "Failed to open dir: %s", strerror(opendir_error));
-        }
+    char * const dir_name = MVM_platform_path(tc, dirname);
+    opendir_error = uv_fs_opendir(NULL, &req, dir_name, NULL);
+    MVM_free(dir_name);
 
-        data->dir_handle = dir_handle;
+    if (opendir_error) {
+        uv_fs_req_cleanup(&req);
+        MVM_exception_throw_adhoc(tc, "Failed to open dir: %s", strerror(opendir_error));
```

Comment:
Maybe `uv_strerror(...)` would give a different error message than `Unknown error` for this:
```
"let's just try the delete loop again"
IO::Path.new("C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\rakudo-precompEN2LSWRQH2TF6XK1L3YXLUJGO\\DD83F90085EDBC48B67BFCC7C5582499ED966F1C\\6B", :SPEC(IO::Spec::Win32), :CWD("D:\\a\\1\\rakudo"))
Failed to get the directory contents of 'C:\Users\VSSADM~1\AppData\Local\Temp\rakudo-precompEN2LSWRQH2TF6XK1L3YXLUJGO\DD83F90085EDBC48B67BFCC7C5582499ED966F1C\6B': Failed to open dir: Unknown error
```

Reply:
The error with `uv_strerror` is 'operation not permitted'. https://dev.azure.com/Rakudo/rakudo/_build/results?buildId=5074&view=logs&j=70f6b8a1-f3cf-5727-bd79-d08ddb1caf67&t=f59a17df-92a9-5d7f-18bb-7f81f1c720a8&l=1948